### ISSUE TITLE
Store images in s3 by camera serial number, then hash

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-base#11]
```
/{serial_number}/
├── /{image_hash}/ 
│   │── original.jpg
│   │── thumbnail.jpg
```

**Comment by postfalk** [comment]
It might be better to store thumbnails in their own structure since they can be considered as helpers to the app opposed to data. 

```
images
├── /{serial_number}/
    ├── {image_hash}.jpg
thumbnails
├── /{serial_number}/
    ├── {image_hash}_small.jpg
```

Personally, I would be even more radical since directories are meaningless on S3 anyways they just get interpreted conventionally. I would not maintain the camera serial number since we might have images where this information is simply not available. I would consider that metadata.

**Comment by nathanielrindlaub** [comment]
OK cool yeah I was torn between the thumbnail as a separate "directory" at the root level or nested within each image folder. Good point about the thumbnail being an app helper though - I'll go that route. 

Initially I was going to go with a flat approach (all images in one jumbo directory, not structured by serial number) as you're describing, because I get that it's pretty much meaningless, but after imagining loading our existing 400k SCI images into the bucket, I thought at least grouping them in directories for individual cameras would provide some order to anyone (me) who may need to look at the bucket and find something. A folder with half a million files in it obviously isn't navigable at all. That said I don't feel particularly strongly about it either way.

We talked about this last call, but I do feel strongly that having a serial number and a created date is going to be a requirement for all images in the app. I need to think of a graceful way to deal with images that don't have that info (maybe they can get stored in some directory in s3 to deal with manually), but they will not get forwarded along to the rest of the image processing and storage pipeline without those metadata. The reason for this is because assigning each image to it's respective deployment (a camera deployed at a location for a period of time) is going to be a critical part of project management, user permissions, and future training jobs.

**Comment by postfalk** [comment]
Yeah, you can do that but then you are relying on some external information that might jeopardize the integrity of your storage at some point. But if you wanna do it most people probably would.

The application should also work with manually uploaded images and some applications might strip EXIF data. The point of hashing the images is not to rely on anything but the actual file. 

I know that is hard for humans but generally the reason why data systems break. If you go with serial numbers than there is no point in hashing at all. Then you can also go with date and original filename. On the other hand, your metadata should allow for constructing views that get you the information you want. That could basically almost look like an S3 bucket dashboard and give you an hierarchical structure of your files. 

That would be basically a distinction between storage (that should be utmost abstract) and a view that would allow for all kinds of semantic orders without jeopardizing storage. A good example for such a structure are Salesforce objects that have a uniform URL no matter whose project it is and what the content is about.

**Comment by nathanielrindlaub** [comment]
If the exif data gets stripped or modified, the hash would be different too, right?

**Comment by postfalk** [comment]
Right, but if the incoming data is already stripped and we don't change it we would have a stable file. I am fine with double track modified images. Since we don't have a reliable means to track those changes we need to have a storage system that is robust to accommodate this. Whether, we deal with it on application in a soft way would be a different question.

I am just advocating for a robust abstraction with minimal assumptions. If you want to put limits on we always could do that on the application level.

**Comment by postfalk** [comment]
Another idea would be to maintain a second S3 bucket with unaltered filenames organized by serial numbers that would backup all the data coming in through this pipeline. Basically a non-operational data archive.

It is very common in such complex applications to work of an operational copy of the data which has very different requirements than an archive.

**Comment by nathanielrindlaub** [comment]
Interesting, I like that idea of the two s3 buckets and using one as an archive.

I don't quite follow the stripping of the incoming image metadata. I don't think that's something we've discussed in the past, but I'm definitely open to it. Based on what you've described I'm imagining a workflow like this:

1) image gets uploaded to the archive s3 bucket to a directory called ```ingest/``` or ```staging/``` or something like that (that's necessary because we won't have the serial numbers until we run them through the lambda function, so we can't store them by serial number right away)

2) the s3 object created event triggers the lambda function, which:
- parses the metadata
- moves the image from the ```staging/``` folder to a different one so the final object path might look like: ```archive/{serial_number}/{original_file_name}.jpg```
- if the image does not have serial number or created date, put it somewhere else like ```purgatory/``` or leave it in the staging directory to be reviewed and dealt with manually
- sends the extracted metadata to the express app for storage in a POST request
- strips out all metadata, hashes the image, then saves the image to the operational s3 bucket using {hash}.jpg as the object name.

Is that what you're thinking?

**Comment by postfalk** [comment]
No I am not talking about stripping metadata, I am talking about receiving images with metadata already stripped. Whether or not strip out metadata is an interesting question.

**Comment by nathanielrindlaub** [comment]
Here is my thinking about requiring the serial number and created date in the app. For starters, I do not think this requirement is a high bar or huge assumption to make. All 403k SCI images have these data. It will be a rare exception when users try to upload images that do not.

In exchange for this requirement, we can allow users to create records on the front end for their new and previous deployments, and all images they upload manually from SD cards or automatically through the wireless cams will get automatically mapped to their deployments. The concept using deployments as a central organizational entity and making it a requirement that each image gets mapped to a single deployment solves a bunch of problems at once:

- **project management, permissions, and routing to ml models** - i.e., how do we determine which images belong to which projects, how do we determine which images individual users have access to, and how do users set and then we route specific sets of incoming images to specific ml model endpoints? Cameras can get reused across projects, so we can’t just key off of the camera serial numbers. A deployment also might be relevant to multiple projects at once (maybe Juli has a biosecurity project that has access to a bunch of deployments but she also wants to share them with Christie’s skunk project). I imagine each project entity will have certain configuration options including what models to run the incoming images against, and maybe some automated notifications to certain users. But in order to do any of that we need to have a project entity that’s associated with some deployment entities, and to be able to route incoming images to their respective deployments automatically.

- **Future training jobs** - because there is so much consistency in the background of images of camera trap images taken during the same deployments, CNNs trained on camera trap images are highly susceptible to over fitting if you don’t split your training and validation datasets out by location. My hunch is that nanolayers did not do this because they did not have location data for the images, so my guess is that MIRA’s accuracy on trans (new) locations is actually quite a bit lower than what they’ve reported by running against their test images. Going forward I'd like to safeguard against this potential pitfall.

The problem is users and camera manufacturers don’t store information about location and deployments in a consistent way. Some camera manufacturers use built in exif fields to store user entered GPS coordinates. Others use user-set text fields in the metadata but these are often not set or set with some inconsistent naming convention. Most often it seems like if there’s any deployment tracking at all it’s happening on some external spreadsheet. The point is by requiring users to create a record of a deployment in the app first (which will be easy, they just need to give it a name, camera serial number, start date, and end date if it’s a deployment in the past... GPS coords would be a big plus but not even a requirement), we will be adding a ton of highly valuable structure to the data at low cost to the user and with pretty reasonable assumptions about the metadata (that the image will have serial number and created date).

If it becomes apparent that not having those metadata it is an issue, one idea might be that for manual uploads, there could be a setting where if the uploader function can’t find a serial number or created date for an image, it could override the automated mapping function and manually assign it to a deployment of a users choosing.

**Comment by nathanielrindlaub** [comment]
Sorry for the long post. Take some time to read through all of that, and let's come up with a solution for the image ingestion and storage together. Happy to jump on a call to discuss; I'd just like to make a decision and implement it.

**Comment by nathanielrindlaub** [comment]
Update on this one - I just finished implemented the following image ingestion workflow. FYI it all is handled by the [lambda](https://github.com/tnc-ca-geo/animl-lambda) function, not the script on the base, so it now pertains to that repo not this one.

I set up three s3 buckets: `animl-data-staging`, `animl-data-archive`, and `animl-data-production`. The staging bucket is where this RPi script uploads new images to and where all other data would be added. Adding a new object to this bucket triggers the lambda, which extracts the metadata, creates a thumbnail, creates a hash, and then copies the image(s) to the production bucket, with a structure of:

```
images/
├── {image_hash}
thumbnails/
├── {image_hash}-small
```

and it copies the images to the archive with a structure of:

```
{serial_number}/
├── {original_file_name}/
         ├── {image_hash}
unknown-camera/
├── {original_file_name}/
         ├── {image_hash}
```

If the serial number is unknown, it will file it in the `unknown-camera` directory, otherwise it will organize the images by camera. The reason I treated the `original_file_name` as a directory one level below the camera directory, and the image hash as the actual object, was to avoid over-writing data if someone uploaded an image with the same serial number and filename as an image already in the archive.

I'm still definitely open feedback and to adjusting all of this; this just was my attempt at synthesizing everything we'd talked about thus far. Closing it out for now though.
